lib/pbench/server/api/resources/visualize.py

```diff
@@ -0,0 +1,98 @@
+from http import HTTPStatus
+from urllib.request import Request
+
+from flask import current_app, jsonify
+from flask.wrappers import Response
+from pquisby.lib.post_processing import BenchmarkName, InputType, QuisbyProcessing
+
+from pbench.server import OperationCode, PbenchServerConfig
+from pbench.server.api.resources import (
+    APIAbort,
+    ApiAuthorizationType,
+    ApiBase,
+    ApiContext,
+    APIInternalError,
+    ApiMethod,
+    ApiParams,
+    ApiSchema,
+    Parameter,
+    ParamType,
+    Schema,
+)
+from pbench.server.cache_manager import (
+    CacheManager,
+    TarballNotFound,
+    TarballUnpackError,
+)
+from pbench.server.database import Dataset
+
+
+class Visualize(ApiBase):
+    """
+    API class to retrieve data using Quisby to visualize
+    """
+
+    def __init__(self, config: PbenchServerConfig):
+        super().__init__(
+            config,
+            ApiSchema(
+                ApiMethod.GET,
+                OperationCode.READ,
+                uri_schema=Schema(
+                    Parameter("dataset", ParamType.DATASET, required=True),
+                ),
+                authorization=ApiAuthorizationType.DATASET,
+            ),
+        )
+
+    def _get(
+        self, params: ApiParams, request: Request, context: ApiContext
+    ) -> Response:
+        """
+        This function is using Quisby to process results into a form that supports visualization
+
+        Args:
+            params: includes the uri parameters, which provide the dataset.
+            request: Original incoming Request object
+            context: API context dictionary
+
+        Raises:
+            APIAbort, reporting "NOT_FOUND" and "INTERNAL_SERVER_ERROR"
+
+        GET /api/v1/visualize/{dataset}
+        """
+
+        dataset = params.uri["dataset"]
+
+        cache_m = CacheManager(self.config, current_app.logger)
+        try:
+            tarball = cache_m.find_dataset(dataset.resource_id)
+        except TarballNotFound as e:
+            raise APIAbort(HTTPStatus.NOT_FOUND, str(e))
+
+        name = Dataset.stem(tarball.tarball_path)
+        try:
+            file = tarball.extract(tarball.tarball_path, f"{name}/result.csv")
+        except TarballUnpackError as e:
+            raise APIInternalError(str(e)) from e
+
+        metadata = self._get_dataset_metadata(
+            dataset, ["dataset.metalog.pbench.script"]
+        )
+        benchmark = metadata["dataset.metalog.pbench.script"]
+        if benchmark == "uperf":
+            benchmark_type = BenchmarkName.UPERF
+        else:
+            raise APIAbort(HTTPStatus.UNSUPPORTED_MEDIA_TYPE, "Unsupported Benchmark")
```

Comment:
I'm inclined to say we can/should be dependent on exactly what Pbench Agent generates; because if we somehow end up with a tarball that has "similar" data (e.g., a `pbench.metalog.pbench.script` value in a different case) then it's *not* from the Pbench Agent wrapper script and hasn't undergone the same Agent-side postprocessing and won't have the expected contents anyway regardless of our "recognition" of a "supported" script.

Ideally, we want to move away from Agent postprocessing entirely, and we'd recognize a "uperf" run because it possesses the directory structure and files we expect from uperf, regardless of the benchmark script name; e.g., including `pbench-user-benchmark -- uperf` as well as `pbench-user-benchmark -- arbitrary_script_name_that_actually_runs_uperf`.

In essence: being "generous" here isn't really doing anyone any favors.

Reply:
> being "generous" here isn't really doing anyone any favors.

OK, but I really feel that the Server shouldn't be involved in this logic (at least, not as long as it is the Agent which is generating the input file for Quisby).  In particular, we shouldn't have to change the Server code when Quisby support expands.

This code should be as close to a pass-through as we can make it:  it shouldn't be up to the Server to decide whether Quisby can handle a given result...that should be up to Quisby!

So, if the Server has to tell Quisby what kind of input file it thinks it is passing, then we should be doing some sort of look-up based on a table that Quisby provides, not an if/else tree in the Server code.  But, it would be better still if we could just pass the input file straight to Quisby and let it decide if it is supported or not.

Reply:
Certainly a list lookup (basically `dataset.metalog.pbench.script in supported_benchmarks`) would be nice. Right now, for whatever internal reasons, this isn't based on or restricted by Quisby (which supports fio plus a bunch of others) but rather on the Pbench Server integration code, which needs to at least grow a list of benchmarks to identify. (And for whatever reason we decided to go with `uperf` first.) But we could certainly have our own list, or a dict if we need to know more about what to look for in each. I wasn't inclined to worry about that for the first singular `uperf` stage, but it'd be better when we add `fio`.

Reply:
> we could certainly have our own list

I'm hung up on who the "we" in that assertion is.  I don't think the Pbench Server code should be maintaining a list of what is or is not available in Quisby.  Yes, we can turn that inside out and claim that the list is what the Pbench Server supports and not what Quisby supports.  However, I think it would be best to avoid the issue altogether, if we can find a way to compartmentalize the decision and push it out of the Server code.  If that's only as far as the "Pbench Server integration code"...well, that's still better.

> I wasn't inclined to worry about that for the first singular `uperf` stage, but it'd be better when we add `fio`.

The problem is that, when most coders see an `if` statement, they are inclined to just add and `elif` to it to cover the new case.  And, as I said, I would like to see us be able to expand to cover `fio` without having to change the Pbench Server code itself -- just pull in an updated Quisby (or the integration code for it) and voila!  So, now _would_ be a good time to consider how we do this.